enhancements/machine-api/azure-ultra-disks.md

```diff
@@ -0,0 +1,261 @@
+This will allow users to set Ultra Disks as Data Disks for the Machine. This feature has been made available upstream in Cluster API Provider for Azure (CAPZ) by: https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1478
+1. Required configuration for specifying the ability in `AzureMachineProviderSpec` to allow the Azure CSI driver to provision Ultra Disks as PVs for the Machine. See the issue tracking the implementation of this feature in the upstream Cluster API Provider for Azure (CAPZ): https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/1852
+
+### User Stories
+
+**Story 1**
+As a developer, I'd like to have the ability to use Azure Ultra Disks as Data Disks for my Machines.
+
+**Story 2**
+As a developer, I'd like to specify some basic specs (Disk Size, Logical Unit Number (LUN), Caching Type) for creating Azure Ultra Disks as Data Disks for my Machines.
+
+**Story 3**
+As a developer, I'd like to have the ability to create Persistent Volume Claims (PVCs) which can dynamically bind to a StorageClass backed by Azure Ultra Disk and to mount them successfully to Pods for my workloads needs.
+
+### API Extensions
+
+#### Extension for Ultra Disks as Data Disks
+
+The API design is taken from the upstream Cluster API Provider for Azure (CAPZ), [here](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/fee74172efc0db07a244ceb870be759db56e0f83/api/v1beta1/azuremachine_types.go#L77-L79) and [here](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/fee74172efc0db07a244ceb870be759db56e0f83/api/v1beta1/types.go#L465-L483).
+The upstream design is followed as closely as possible to ensure smoother future migrations between MAPI and CAPI.
+
+A new field will be added to the `AzureMachineProviderSpec` struct `DataDisks`. The interface will be a slice of `DataDisk` structs allowing multiple Data Disks and various fields specific to them to be further specified.
+
+A `DataDisk` has one Composite type field, `ManagedDisk` of type `ManagedDiskParameters`, which is a type already defined in the Machine API spec, used for the `ManagedDisk` field in the `OSDisk` type struct, another `AzureMachineProviderSpec` field.
+
+
+```go
+type AzureMachineProviderSpec struct {
+  // Existing fields will not be modified
+  ...
+
+  // DataDisk specifies the parameters that are used to add one or more data disks to the machine
+  // +optional
+  DataDisks []DataDisk `json:"dataDisks,omitempty"`
+}
+
+// DataDisk specifies the parameters that are used to add one or more data disks to the machine.
+type DataDisk struct {
+  // NameSuffix is the suffix to be appended to the machine name to generate the disk name.
+  // Each disk name will be in format <machineName>_<nameSuffix>.
+  // The overall disk name must not exceed 80 chars in length and can only contain letters, numbers, underscores, periods or hyphens.
+  NameSuffix string `json:"nameSuffix"`
```

Comment:
In terms of kubebuilder tags we could add here, the length is a hard one though as while we know the maximum is 80, we can't suggest a smaller maximum than 78 (1 char machine name plus underscore to join)
```suggestion
  // +kubebuilder:validation:Pattern:="[a-zA-Z0-9][\w.-]+[a-zA-Z0-9]"
  // +kubebuilder:validation:MaxLength:=78
  // +kubebuilder:validation:Required
  NameSuffix string `json:"nameSuffix"`
```

Reply:
Even though we don't need the restriction of an alphanumeric starting character on the suffix, given we are adding a prefix and an underscore, I think we should enforce that rule anyway.

Image for example someone sets a name suffix of `-foo`, now the generate name might be `machine_-foo`, which is ugly.

We don't have to enforce this, but IMO it will make it nicer from a UX perspective. We are able to enforce this if we so desire to do so

Reply:
After a chat offline we agreed to go with name suffixes that start and end with an alphanumeric character, while containing the chars specified by _3._ for UX reasons.